### PR TITLE
Adding distributed transaction support

### PR DESCRIPTION
**Pull request by VijithaEkanayake**

### CODE REVIEW

File: modules/jms/pom.xml [hunk 1]
```diff
@@ -154,6 +154,11 @@
             <artifactId>junit</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.atomikos.wso2</groupId>
+            <artifactId>atomikos</artifactId>
+            <version>3.8.0.wso2v1</version>
```
Comment (Buddhima) [inline modules/jms/pom.xml]: version is not allowed in this level pom file. Add dependency to root pom file -> dependancyManagement. Only groupId & artifactId should be mentioned here.

File: modules/jms/src/main/java/org/apache/axis2/transport/jms/JMSMessageSender.java [hunk 2]
```diff
@@ -155,12 +187,11 @@ public void send(Message message, MessageContext msgCtx) {
             } else {
                 if (isQueue) {
                     try {
-                        ((QueueSender) producer).send(message);
+                        (producer).send(message);
```
Comment (Buddhima) [inline modules/jms/src/main/java/org/apache/axis2/transport/jms/JMSMessageSender.java]: brackets are not needed

File: modules/jms/src/main/java/org/apache/axis2/transport/jms/JMSMessageSender.java [hunk 3]
```diff
@@ -155,12 +187,11 @@ public void send(Message message, MessageContext msgCtx) {
...
                     } catch (JMSException e) {
                         createTempQueueConsumer();
-                        ((QueueSender) producer).send(message);
+                        (producer).send(message);
```
Comment (Buddhima) [inline modules/jms/src/main/java/org/apache/axis2/transport/jms/JMSMessageSender.java]: same as above

File: modules/jms/src/main/java/org/apache/axis2/transport/jms/JMSSender.java [hunk 4]
```diff
@@ -124,8 +132,27 @@ public void sendMessage(MessageContext msgCtx, String targetAddress,
...
+                    } else {
+                        ArrayList list = JMSMessageSenderMap.get(transaction);
+                        list.add(messageSender);
+                        JMSMessageSenderMap.put(transaction, list);
+                    }
+
                 } catch (JMSException e) {
+                    Transaction transaction = null;
+                    try {
+                        transaction = ((TransactionManager) msgCtx.getProperty("distributedTxMgr")).getTransaction();
```
Comment (Buddhima) [inline modules/jms/src/main/java/org/apache/axis2/transport/jms/JMSSender.java]: use the constant defined for "distributedTxMgr"

File: modules/jms/src/main/java/org/apache/axis2/transport/jms/JMSSender.java [hunk 5]
```diff
@@ -186,6 +215,13 @@ private void sendOverJMS(MessageContext msgCtx, JMSMessageSender messageSender,
         try {
             message = createJMSMessage(msgCtx, messageSender.getSession(), contentTypeProperty);
         } catch (JMSException e) {
+            Transaction transaction = null;
+            try {
+                transaction = ((TransactionManager) msgCtx.getProperty("distributedTxMgr")).getTransaction();
```
Comment (Buddhima) [inline modules/jms/src/main/java/org/apache/axis2/transport/jms/JMSSender.java]: use the constant defined for "distributedTxMgr" - for all occurrences

File: modules/jms/src/main/java/org/apache/axis2/transport/jms/JMSOutTransportInfo.java [hunk 6]
```diff
@@ -354,53 +364,120 @@ public JMSMessageSender createJMSSender() throws JMSException {
...
+            Xid xid1 = null;
+            Transaction transaction = null;
+
+            try {
+                tx = (TransactionManager) msgCtx.getProperty(JMSConstants.JMS_XA_TRANSACTION_MANAGER);
+                transaction = tx.getTransaction();
+                msgCtx.setProperty(JMSConstants.JMS_XA_TRANSACTION_MANAGER, tx);
+                msgCtx.setProperty(JMSConstants.JMS_XA_TRANSACTION, transaction);
+                xid1 = new JMSXid(JMSConstants.JMS_XA_TRANSACTION_PREFIX.getBytes(StandardCharsets.UTF_8), 1,
+                        UUIDGenerator.getInstance().generateStringUUID().getBytes());
```
Comment (Buddhima) [inline modules/jms/src/main/java/org/apache/axis2/transport/jms/JMSOutTransportInfo.java]: Isn't it possible to re-use a UUIDGenerator, rather including UUIDGenerator related code?

File: modules/jms/src/main/java/org/apache/axis2/transport/jms/UUIDGenerator.java [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+    public byte[] generateDummyAddress() {
+        Random rnd = getRandomNumberGenerator();
+        byte[] dummy = new byte[6];
+        rnd.nextBytes(dummy);
+      /* Need to set the broadcast bit to indicate it's not a real
+       * address.
+       */
+        dummy[0] |= (byte) 0x01;
+
+//        if (ActiveMQUtilLogger.LOGGER.isDebugEnabled()) {
```
Comment (virajsenevirathne) [inline modules/jms/src/main/java/org/apache/axis2/transport/jms/UUIDGenerator.java]: Please remove those

File: modules/jms/src/main/java/org/apache/axis2/transport/jms/JMSOutTransportInfo.java [hunk 8]
```diff
@@ -354,53 +362,121 @@ public JMSMessageSender createJMSSender() throws JMSException {
...
-        
-        Connection connection = null;
-        if (user != null && pass != null) {
-            if (qConFac != null) {
-                connection = qConFac.createQueueConnection(user, pass);
-            } else if (tConFac != null) {
-                connection = tConFac.createTopicConnection(user, pass);
+        }
+
+        if(msgCtx.getProperty(JMSConstants.JMS_XA_TRANSACTION_MANAGER) != null) {
```
Comment (virajsenevirathne) [inline modules/jms/src/main/java/org/apache/axis2/transport/jms/JMSOutTransportInfo.java]: formattin issue

**Pull request merged**